currency/storage.go

```diff
@@ -48,142 +78,132 @@ func (s *Storage) SetDefaults() {
+	if settings.ForeignExchangeUpdateDuration <= 0 {
+		return errInvalidForeignExchangeUpdateDuration
 	}
 
-	if settings.FxRateDelay.Nanoseconds() == 0 {
-		s.foreignExchangeUpdateDelay = DefaultForeignExchangeDelay
+	s.mtx.Lock()
+	s.shutdown = make(chan struct{})
+	s.baseCurrency = settings.FiatDisplayCurrency
+	s.path = filepath.Join(filePath, DefaultStorageFile)
+	s.currencyFileUpdateDelay = settings.CurrencyFileUpdateDuration
+	s.foreignExchangeUpdateDelay = settings.ForeignExchangeUpdateDuration
+
+	log.Debugf(log.Currency, "Fiat display currency: %s.\n", s.baseCurrency)
+	var err error
+	if overrides.CurrencyConverter &&
+		settings.CryptocurrencyProvider.APIKey != "" &&
+		settings.CryptocurrencyProvider.APIKey != "Key" {
+		log.Debugln(log.Currency, "Setting up currency analysis system with Coinmarketcap...")
+		s.currencyAnalysis, err = coinmarketcap.NewFromSettings(coinmarketcap.Settings(settings.CryptocurrencyProvider))
+		if err != nil {
+			log.Errorf(log.Currency, "Unable to setup CoinMarketCap analysis. Error: %s", err)
+		}
 	} else {
-		s.foreignExchangeUpdateDelay = settings.FxRateDelay
+		log.Warnf(log.Currency, "%s API key not set, disabling. Please set this in your config.json file\n",
+			settings.CryptocurrencyProvider.Name)
 	}
 
 	var fxSettings []base.Settings
+	var primaryProvider bool
 	for i := range settings.ForexProviders {
-		switch settings.ForexProviders[i].Name {
-		case "CurrencyConverter":
-			if overrides.FxCurrencyConverter ||
-				settings.ForexProviders[i].Enabled {
-				settings.ForexProviders[i].Enabled = true
-				fxSettings = append(fxSettings,
-					base.Settings(settings.ForexProviders[i]))
-			}
-
-		case "CurrencyLayer":
-			if overrides.FxCurrencyLayer || settings.ForexProviders[i].Enabled {
-				settings.ForexProviders[i].Enabled = true
-				fxSettings = append(fxSettings,
-					base.Settings(settings.ForexProviders[i]))
-			}
+		enabled := (settings.ForexProviders[i].Name == "CurrencyConverter" && overrides.CurrencyConverter) ||
+			(settings.ForexProviders[i].Name == "CurrencyLayer" && overrides.CurrencyLayer) ||
+			(settings.ForexProviders[i].Name == "Fixer" && overrides.Fixer) ||
+			(settings.ForexProviders[i].Name == "OpenExchangeRates" && overrides.OpenExchangeRates) ||
+			(settings.ForexProviders[i].Name == "ExchangeRates" && overrides.ExchangeRates) ||
+			(settings.ForexProviders[i].Name == "ExchangeRateHost" && overrides.ExchangeRateHost)
+
+		if !enabled {
+			continue
+		}
 
-		case "Fixer":
-			if overrides.FxFixer || settings.ForexProviders[i].Enabled {
-				settings.ForexProviders[i].Enabled = true
-				fxSettings = append(fxSettings,
-					base.Settings(settings.ForexProviders[i]))
+		if settings.ForexProviders[i].Name != DefaultForexProviderExchangeRatesAPI {
+			if settings.ForexProviders[i].APIKey == "" || settings.ForexProviders[i].APIKey == "Key" {
+				log.Warnf(log.Currency, "%s forex provider API key not set, disabling. Please set this in your config.json file\n",
```

Comment:
This is an inaccurate error. My cyrptocurrencyprovider "lksadf" has an API key set

Reply:
Can you send me the log output for this please, it seems to be working for me?

Reply:
The log won't matter much as that's the only really relevant line.
What you might want is the config:
```
  "cryptocurrencyProvider": {
   "name": "lkjkjhl",
   "enabled": true,
   "verbose": true,
   "apiKey": "1234-5678",
   "accountPlan": "Basic"
  },
```